lib/syntaxtree.gi

```diff
@@ -0,0 +1,831 @@
+    # can contain arbitrary expressions
+    compiler.T_PERM_EXPR := function(expr)
+        return rec( type := expr.type
+                  , cycles := List( expr.cycles
+                                  , cyc -> List(cyc, compile) ) );
+    end;
+
+    compiler.T_PERM_CYCLE := function(expr)
+        Error("encountered T_PERM_CYCLE");
+    end;
+
+    compiler.T_LIST_EXPR := function(expr)
+        return rec( type := expr.type
+                  , list := List(expr.list, compile));
+    end;
+
+    compiler.T_LIST_TILD_EXPR := function(expr)
+    end;
+
+    compiler.T_RANGE_EXPR := function(expr)
+        return expr;
+    end;
+
+    compiler.T_REC_EXPR := function(expr)
+        local kvcomp;
+        kvcomp := function(r)
+            local res;
+            res := rec();
+            if IsString(r.key) then
+                res.key := r.key;
+            else
+                res.key := compiler.compile(r.key);
+            fi;
+            res.value := compiler.compile(r.value);
+            return res;
+        end;
+
+        return rec( type := expr.type
+                  , keyvalue := List(expr.keyvalue, kvcomp) );
+    end;
+
+    compiler.T_REC_TILD_EXPR := function(expr) end;
+
+    # Different variable references, and the
+    # appropriate IsBound constructs
+    compiler.T_REFLVAR := expr -> expr;
+    compiler.T_ISB_LVAR := expr -> expr;
+    compiler.T_REF_HVAR := expr -> expr;
+    compiler.T_ISB_HVAR := expr -> expr;
+    compiler.T_REF_GVAR := expr -> expr;
+    compiler.T_ISB_GVAR := expr -> expr;
+
+    compiler.T_ELM_LIST := compile_record;
+    compiler.T_ELMS_LIST := compile_record;
+    compiler.T_ELM_LIST_LEV := compile_record;
+    compiler.T_ELMS_LIST_LEV := compile_record;
+    compiler.T_ISB_LIST := compile_record;
+    compiler.T_ELM_REC_NAME := compile_record;
+    compiler.T_ELM_REC_EXPR := compile_record;
+    compiler.T_ISB_REC_NAME := compile_record;
+    compiler.T_ISB_REC_EXPR := compile_record;
+    compiler.T_ELM_POSOBJ := compile_record;
+    compiler.T_ELMS_POSOBJ := compile_record;
+    compiler.T_ELM_POSOBJ_LEV := compile_record;
+    compiler.T_ELMS_POSOBJ_LEV := compile_record;
+    compiler.T_ISB_POSOBJ := compile_record;
+    compiler.T_ELM_COMOBJ_NAME := compile_record;
+    compiler.T_ELM_COMOBJ_EXPR := compile_record;
+    compiler.T_ISB_COMOBJ_NAME := compile_record;
+    compiler.T_ISB_COMOBJ_EXPR := compile_record;
+    compiler.T_FUNCCALL_OPTS := compile_record;
+    compiler.T_ELM2_LIST := compile_record;
+    compiler.T_ELMX_LIST := compile_record;
+    compiler.T_ASS2_LIST := compile_record;
+    compiler.T_ASSX_LIST := compile_record;
+
+    return Objectify(GAPCompilerType, rec( name := "CleanupCompiler"
+                                         , compiler := compiler) );
+end);
+
+# Pretty printer based on Wadler's Prettier Printer.
+# It is hugely inefficient I guess, but fun.
+# Also this only works on a "cleaned" syntax tree (probably flag this somehow)
```

Comment:
It'd be really nice to have an example for this in a .tst or as a GAPDoc example, or even in a plain source comment here :-).